### PR TITLE
Update build process for flexibility and torch version compatibility

### PR DESCRIPTION
**Pull request by wasertech**
This pull request updates the build process to be more flexible with the required versions of `CC` and `CXX` environment variables. It also allows for the use of different versions of PyTorch by extracting the version from the environment variable `TORCH_VERSION` or the installed one. This ensures compatibility with different PyTorch versions and provides more flexibility in the build process.

## Why is this needed?

To make it easier to build on Arch/Manjaro Linux (which have bleeding edge versions of CUDA, GCC and PyTorch) without having to modify `setup.py`.

## Example usage

On arch you can build from source (in a venv for instance) like so without editing `setup.py`:

```shell
# i.e with Python 3.12 and CUDA 12.5
pacman -S python cuda

# Dependencies to build a venv
pacman -S python-virtualenv

# Clone this branch locally
git clone https://github.com/wasertech/AutoAWQ_kernels.git
cd AutoAWQ_kernels
git checkout -b py312-torch240-cu124

# Setup a virtual env.
python -m venv .venv
source .venv/bin/activate

# Install numpy and torch+cu124
pip install -U numpy
pip install -U torch --index-url https://download.pytorch.org/whl/cu124

# Build the kernels
CC="g++-13" CXX="g++-13" pip install --no-build-isolation -e .
```
Checkout the logs [here](https://gist.github.com/wasertech/ff2cbb8ecaadaeedcabd50d7a06c64f3).

Or in one line:
> [!CAUTION]
> Requires `torch` and `numpy` already installed!
```
CC="${GPP}" CXX="${GPP}" pip install --no-build-isolation "git+https://github.com/wasertech/AutoAWQ_kernels.git@py312-torch240-cu124"
```
Where `$GPP` is the name executable of `g++` as found in the `$PATH`. Ideally the same version that was used to build `torch` (but can be as close as possible). It needs to be compatible with CUDA (<= 13 for CUDA 12.5 and below).

See the [Host Compiler Support Policy](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#host-compiler-support-policy) for your installed version of CUDA.

I think with CUDA 12.6 and above this could be a lot easier in the future but it's still a nice addition to have choice when building from source.

PS: I tried to keep the same default behaviour so to not impact other systems.

### CODE REVIEW

**Comment by wasertech** [comment]
Without those mods, you'll get an error build from source as stated in my comment [here](https://github.com/casper-hansen/AutoAWQ_kernels/pull/25#issuecomment-2266745525)

**Comment by wasertech** [comment]
Once the kernels are built and installed, you need to install AWQ from my branch (https://github.com/casper-hansen/AutoAWQ/pull/573) otherwise it would downgrade pytorch again.

File: setup.py [hunk 1]
```diff
@@ -56,7 +59,7 @@
 }
 
 requirements = [
-    "torch==2.3.1",
+    f"torch=={TORCH_VERSION}",
```
Comment (suparious) [inline setup.py:62]: I'd really prefer to see this get unlocked with a `>-` instead of `==`.

**Comment by stenreijers** [comment]
Any ETA when this gets merged @casper-hansen?

File: setup.py (hunk 1 above)
Comment (wasertech) [reply setup.py:62]: It's acceptable. I'll make the changes here and in the AWQ package too once I've synced it with upstream.

Comment (wasertech) [reply setup.py]: https://github.com/casper-hansen/AutoAWQ_kernels/pull/29/commits/c3a922fdb070240fa5e602155f130e573a38120a should do just fine.

File: setup.py [hunk 2]
```diff
@@ -5,10 +5,14 @@
...
 
-os.environ["CC"] = "g++"
-os.environ["CXX"] = "g++"
+if "CC" not in os.environ:
+    os.environ["CC"] = "g++"
+if "CXX" not in os.environ:
+    os.environ["CXX"] = "g++"
 AUTOAWQ_KERNELS_VERSION = "0.0.7"
 PYPI_BUILD = os.getenv("PYPI_BUILD", "0") == "1"
+COMPUTE_CAPABILITIES = os.getenv("COMPUTE_CAPABILITIES", "75,80,86,87,89,90")
```
Comment (wasertech) [inline setup.py:14]: I've also added a way to to define more computing architectures without editing the file.
PS: I merged #17 to add support for Jetson Orin devices (`87`)

File: setup.py [hunk 3]
```diff
@@ -90,7 +94,7 @@ def get_generator_flag():
 
 
 def get_compute_capabilities(
-    compute_capabilities={75, 80, 86, 89, 90}
+    compute_capabilities=set(map(int, COMPUTE_CAPABILITIES.split(",")))
```
Comment (wasertech) [inline setup.py:97]:
```python
>>> import os
>>> COMPUTE_CAPABILITIES = os.getenv("COMPUTE_CAPABILITIES", "75,80,86,87,89,90")
>>> set(map(int, COMPUTE_CAPABILITIES.split(",")))
{75, 80, 86, 87, 89, 90}
```
Should always create a set of unique integer numbers from `$COMPUTE_CAPABILITIES` (even if it contains doubles).

**Comment by casper-hansen** [comment]
Hi @wasertech, I think this is an excellent PR. I'm sorry for taking so long to get back to you.

I am currently looking at releasing the next version of kernels compiled with torch 2.4.0 and Python 3.12.

Would you mind adding some examples of how to use this in the build from source part in the README?
https://github.com/casper-hansen/AutoAWQ_kernels?tab=readme-ov-file#build-from-source

**Comment by wasertech** [comment]
Hey @casper-hansen ! I'm glad you like it. So I made a first draft to expand the README and tested the instructions for CUDA but I don't have an AMD GPU so I don't really have the way to make sure it works. If someone can test the instruction to build with ROCm and let me know, I would be grateful.

Feel free to review the instructions if something is not clear or you can provide more information.

[Here are the logs](https://gist.github.com/wasertech/0ca238176ba3ceb1ef4603ce2f0783d2). Keep in mind I built from main upstream so it downgraded everything but it build no problem otherwise.

**Pull request merged**